raiden/transfer/mediated_transfer/initiator_manager.py

```diff
@@ -251,6 +253,47 @@ def handle_transferrefundcancelroute(
     return iteration
 
 
+def handle_lock_expired(
+        payment_state: InitiatorPaymentState,
+        state_change: ReceiveSecretReveal,
+        channelidentifiers_to_channels: typing.ChannelMap,
+        pseudo_random_generator: random.Random,
+        block_number: typing.BlockNumber,
+) -> TransitionResult:
+    """Initiator also needs to handle LockExpired messages when refund transfers are involved.
+
+    A -> B -> C
+
+    - A sends locked transfer to B
+    - B attempted to forward to C but has not enough capacity
+    - B sends a refund transfer with the same secrethash back to A
+    - When the lock expires B will also send a LockExpired message to A
+    - A needs to be able to properly process it
+
+    Related issue: https://github.com/raiden-network/raiden/issues/3183
+"""
+    channel_identifier = payment_state.initiator.channel_identifier
+    channel_state = channelidentifiers_to_channels[channel_identifier]
+    secrethash = payment_state.initiator.transfer.lock.secrethash
+    result = channel.handle_receive_lock_expired(
+        channel_state=channel_state,
+        state_change=state_change,
+        block_number=block_number,
+    )
+
+    if not channel.get_lock(result.new_state.partner_state, secrethash):
+        transfer = payment_state.initiator.transfer
+        unlock_failed = EventUnlockClaimFailed(
+            identifier=transfer.payment_identifier,
+            secrethash=transfer.lock.secrethash,
+            reason='Lock expired',
+        )
+        result.events.append(unlock_failed)
+        return TransitionResult(None, result.events)
```

Comment:
Here, you can use `return TransitionResult(payment_state if channel.lock_exists(secrethash) else None, result.events)` if you want to follow my suggestion.

Reply:
@rakanalh I don't follow the scenario you described. If we have two different locked transfers we also will have 2 different payment tasks.

Reply:
Ah correct... i tried to think of ways this could lead to us handling the same issue of a task being cleared while we still need it. The question is: how can this break? the answer would be, in this case, having 2 transfers (1 normal, 1 refund). So if the refund expires, doesn't mean that we should clear the task, but rather wait for the original initiated transfer (in our_state) to be expired to clear the task. So my suggestion here is to only clear the task (in the context of lock expiry) in the handle_block.

Reply:
So I ended up going with removing the task in the next block at which both our state and partner state no longer have a lock with the secrethash of the transfer. I think this is also what you suggested if I understood correctly.